kong/plugins/basic-auth/daos.lua

```diff
@@ -26,4 +27,20 @@ function BasicAuthCredentials:new(properties)
   BasicAuthCredentials.super.new(self, properties)
 end
 
+function BasicAuthCredentials:insert(params)
+  if params.password then
+    return BasicAuthCredentials.super.insert(self, params)
+  else
+    -- No password was provided, so we insert a random generated password
+    local newpwd = utils.random_string()
+    params.password = newpwd
+    local data, err = BasicAuthCredentials.super.insert(self, params)
+    -- inserting the data has encrypted the password field by now, 
+    -- so add a new field with the generated plain text password
+    -- which will be returned to the requester
+    if data then data.plain_password = newpwd end
+    return data, err
+  end
+end
```

Comment:
I'm not following you. In this case we need code executing before the update (generating password), and after (adding it in plaintext). `func` only runs the 'before' code.

Moving the code out of `insert` in a function above, would mean splitting it up in before and after functions. Calling both, and pass the password around in the arguments. The code would become worse imo. Is this what you meant or didn't I understand it correctly?

Looking at the schema definition I think `func` could be renamed to `before_update` and an `after_update` could be added. That would make it more clear imo, and easier to implement this type of updates.
Performance wise you might want to drop them and add the same two on the schema-level instead of the field-level (fewer check and function calls). Maybe could replace the marshall functions on the DAO also, though I haven't really looked in to those.